tests/unit/imutils/test_metadata.py

```diff
@@ -0,0 +1,335 @@
+    metadata_handler, metadata_content, metadata_test_keys
+):
+    for current_keys in metadata_test_keys:
+        valid_keys = [key for key in current_keys if key in metadata_content.keys()]
+        fetched = metadata_handler.fetch_keys(current_keys)
+        assert len(fetched) == len(valid_keys)
+
+        for current_key in valid_keys:
+            assert current_key in fetched
+            value_match(metadata_content[current_key], fetched[current_key][1])
+
+
+def test_metadatahandler_fetch_keys_piexif(
+    metadata_handler_piexif, metadata_content, metadata_test_keys
+):
+    for current_keys in metadata_test_keys:
+
+        valid_keys = [
+            key
+            for key in current_keys
+            if key in metadata_content.keys() and not ("Iptc" in key or "Xmp" in key)
+        ]
+        print(valid_keys, current_keys)
+        fetched = metadata_handler_piexif.fetch_keys(current_keys)
+        assert len(fetched) == len(valid_keys)
+
+        for current_key in valid_keys:
+            short_key = current_key.rpartition(".")[-1]
+            assert current_key in fetched or short_key in fetched
+            # Internal keys are still in long form
+            try:
+                value_match_piexif(metadata_content[current_key], fetched[short_key][1])
+            except KeyError:
+                value_match_piexif(
+                    metadata_content[current_key], fetched[current_key][1]
+                )
+
+
+def test_metadatahandler_get_keys(
+    metadata_handler, metadata_handler_piexif, metadata_content
+):
+    available_keys = list(metadata_content.keys())
+
+    for handler in (metadata_handler, metadata_handler_piexif):
+        if isinstance(handler._external_handler, metadata._ExternalKeyHandlerPiexif):
+            available_keys = [
+                e for e in available_keys if not ("Iptc" in e or "Xmp" in e)
+            ]
+
+        fetched_keys = list(handler.get_keys())
+
+        # Available_keys does not contain all possible keys
+        assert len(fetched_keys) >= len(available_keys)
+
+        for key in available_keys:
+            short_key = key.rpartition(".")[-1]
+            assert key in fetched_keys or short_key in fetched_keys
+
+
+@pytest.fixture
+def external_keyhandler(add_metadata_information, dummy_image, external_content):
+    assert pyexiv2 is not None, "pyexiv2 required to add metadata information"
+    image = dummy_image()
+    add_metadata_information(image, external_content)
+    return metadata.ExternalKeyHandler(image)
+
+
+@pytest.fixture
+def external_keyhandler_piexif(add_metadata_information, dummy_image, external_content):
+    assert pyexiv2 is not None, "pyexiv2 required to add metadata information"
+    image = dummy_image()
+    add_metadata_information(image, external_content)
+    return metadata._ExternalKeyHandlerPiexif(image)
+
+
+def test_external_keyhandler_get_date_time(
+    external_keyhandler, external_keyhandler_piexif, dummy_image
+):
```

Comment:
This is a really funky way to do this as pytest provides parametrization precisely for this purpose :blush:

As an example, the (unused?) fixture at the top called `external_handler` will make tests run twice, once with `metadata.ExternalKeyHandler` as value for `external_handler` and once with `metadata._ExternalKeyHandlerPiexif` as the argument. This fixture could easily be modified to actually create the handler with some image as path, but it might be nice to do this within this test, so the structure would be something along:
- add metadata to dummy image if any
- create handler with path to dummy image
- compare to expected result

Both the metadata and the expected results could also be parameters, now not of the fixture but of the test itself. The structure would end up looking somewhat like this (dummy code):
```python
@pytest.mark.parametrize("metadata, expected_date", [({"name": ExifTag("name", "value"}, "value"), ({}, "")]
def test_external_keyhandler_copy_metadata(external_handler, dummy_image, add_metadata_information, metadata, expected_date):
    src = dummy_image("image.jpg")
    add_metadata_information(src, metadata)
    handler = external_handler(src)
    assert handler.get_date_time() == expected_date
```

While this doesn't look much better than the four assert statements in your implementation, it does have a few advantages:

- Each assert is run in its own test, i.e. we test pyexiv2 and piexif as well as content / no content separately
- We can very easily add more combinations

This kind of logic applies to some more parts of the tests you wrote, e.g. the copy_metadata test could then also be parametrized instead of having a for loop. Hope this makes sense :blush: